### ISSUE TITLE
[D9][UX] Remove the weight field from the menu item edit form and the node edit form

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#4466]
This is the respective issue for https://www.drupal.org/project/drupal/issues/491022

- The interface is simpler.
- The functionality provided is not very helpful. The majority of users have no idea of the weight assigned to other menu items. This is a frustrating experience for users and a definite "DrupalWTF". Drag and drop menus are an intuitive experience.
- Weight has been removed from other user facing systems. Are there any other areas of core that still show a weight selector like this?
- The change is consistent with blocks. You can not set the weight of a block under its configuration page.

**Comment by klonos** [comment]
PR up for review: https://github.com/backdrop/backdrop-issues/issues/4466

I would eventually like us to move to an intuitive solution like #3029, but this is a step towards less frustrating UX.

**Comment by indigoxela** [comment]
The menu item positioning thing (UI) is something that annoyed me for ages, I'm glad @klonos takes care of it.

But simply removing the weight item from form still doesn't seem ideal to me. Then again one creates the item and has to go to the listing page to position the menu item (clicking around).

One possible improvement comes to my mind: currently - if not set - the default weight is `0`.
It could be a _tiny_ improvement if the new item gets added to the end of the list (highest existing weight + 1 in that level). Not sure how complicated that would be. At least for me adding a menu item at the end of a list is the most common use case.

Other thoughts?

**Comment by indigoxela** [comment]
Here's a possible alternative to evaluate: it turns the node form item to a tabledrag widget:

https://www.drupal.org/project/menu_link_weight

Is anyone using this with Drupal? Here's what it looks like on a vanilla D7:

![menu-item-tabledrag](https://user-images.githubusercontent.com/842066/88374273-c52d6b80-cd99-11ea-97a4-b21eaa5fa279.png)

**Comment by klonos** [comment]
@indigoxela menu_link_weight is one of the modules I've evaluated for #3029 ...that module replaces the menu selector in the node edit form, but it doesn't do anything for the menu item edit form (the 8.x version does - not 7.x).

**Comment by indigoxela** [comment]
Ah, thanks a lot for pointing me to the proper (and old!) Issue. :+1:

**Comment by olafgrabienski** [comment]
> simply removing the weight item from form still doesn't seem ideal to me. Then again one creates the item and has to go to the listing page to position the menu item (clicking around).
>
> One possible improvement comes to my mind: currently - if not set - the default weight is `0`.
It could be a _tiny_ improvement if the new item gets added to the end of the list (highest existing weight + 1 in that level).

@indigoxela The idea for a new default behavior -- new items get added to the end of the list -- sounds helpful for some use cases but for others maybe not. In my understanding it makes sense when you're adding one ore more content items and plan to rearrange their menu links on the menu item list shortly after. In contrast, I've quite often the use case of longer sub-menus where the items are to be listed alphabetically by default, which is happening at the moment with a default weight of `0`.

**Comment by indigoxela** [comment]
>  I've quite often the use case of longer sub-menus where the items are to be listed alphabetically by default

@olafgrabienski I see, then the current behavior is the only option for you.

Can you please post your use case on the [other issue](https://github.com/backdrop/backdrop-issues/issues/3029)? Because if we added that feature to core, it might be a breaking change for you.

**Comment by olafgrabienski** [comment]
> post your use case on the other issue? Because if we added that feature to core, it might be a breaking change for you.

Not sure ... would that feature imply a change of the default weight `0`?

**Comment by indigoxela** [comment]
> would that feature imply a change of the default weight?

That's not clear yet, no one works on that issue currently, but your use case could help to avoid pitfalls.

**Comment by klonos** [comment]
> would that feature imply a change of the default weight `0`?
>
>>That's not clear yet,...

As far as this PR here goes, this functionality is not lost. If you try adding a new piece of content in the PR sandbox, you'll see that by default it is added with weight `0` 😉

**Comment by olafgrabienski** [comment]
> If you try adding a new piece of content in the PR sandbox, you'll see that by default it is added with weight `0`

Great! Tested in the sandbox, works as expected: no weight field in the menu tab of the content edit form, and new menu items default to a weight of `0`.

**Comment by jenlampton** [comment]
@klonos I love this and I think its an improvement, but I do have some concerns about removing the weight field from the menu edit form. I've gotten my sites into some sticky situations where the drag-and-drop is not sufficient for moving the menu links where I want them, and setting the weight manually is my only way back to sanity.

The node edit form is a no-brainer because most editors won't know what weight is anyway, but the menu UI is a useful tool for admins to untangle things, and I don't want to limit their ability to do so. 

The PR looks straightforward and I'm tempted to mark this RTBC, but I just wanted to drop that concern here.

**Comment by klonos** [comment]
> I've gotten my sites into some sticky situations where the drag-and-drop is not sufficient for moving the menu links where I want them, and setting the weight manually is my only way back to sanity.

Valid concern, but in that case, can't you use the "show row weights" feature instead of drag-n-drop? 🤔

Do you recall what caused these sticky situations? Is it a bug or browser incompatibility thing that people may come across frequently, or simply too complex menus, with a huge amount of links to manage? If the later, then https://github.com/backdrop-contrib/bigmenu is a huge help - we've added that module to the GovCMS distro, and customers were really happy with the solution.

**Comment by quicksketch** [comment]
> Do you recall what caused these sticky situations?

When ordering Taxonomy terms, only 100 per page are shown (by default), making it so that if you want to move an item from position 500 to position 1, it's not possible with drag-and-drop. Menus appear to work differently in that they always display _all_ items. But I can just see this issue being the gateway to suggesting the Weight field should be removed everywhere.

**Comment by BWPanda** [comment]
@klonos Can you please rebase the PR so we can get updated test and sandbox running?

**Comment by klonos** [comment]
Sure @BWPanda ...done! 😉

**Comment by BWPanda** [comment]
Thanks @klonos, I was able to see what this issue's referring to better now. Looks good and works for me. However...

> I can just see this issue being the gateway to suggesting the Weight field should be removed everywhere.

Why not just do that here then? Why leave weight fields in other places (e.g. Taxonomy terms) when removing them from menus and nodes?

**Comment by indigoxela** [comment]
FTR, I did a quick check, if this interferes in any way with the Menu Link Weight module, but all still works as expected. :+1:

**Comment by klonos** [comment]
> Why not just do that here then? Why leave weight fields in other places (e.g. Taxonomy terms) when removing them from menus and nodes?

@BWPanda I believe that @quicksketch outlined it perfectly:
> When ordering Taxonomy terms, only 100 per page are shown (by default), making it so that if you want to move an item from position 500 to position 1, it's not possible with drag-and-drop. Menus appear to work differently in that they always display all items, so it doesn't share this exact problem.

**Comment by BWPanda** [comment]
@klonos But you'd still be able to manually edit a term's weight by enabling the weight fields on the listing page. So I see no reason the weight field can't be removed from the individual term's Edit pages...

**Comment by laryn** [comment]
I tend to agree that the weight field could be removed on the individual taxonomy term pages, too, but was going to get this change in not to hold things up, since it's been sitting around for a while already. Unfortunately when I rebased the tests came back with some failures that I don't have time to dig into right now... so back to needs work.

If I can summarize the thread as I understand it:
 - removing these "weight" fields from individual creation pages (which have no context about other element's weights) is a UX win
 - we need to make sure weight fields still exist as an option on the summary pages where multiple elements are listed in relation to each other

**Comment by stpaultim** [comment]
Is this relevent? (I'm not sure if it is, but this issue came up in dev meeting on the same day as the forum discussion is happening).

https://forum.backdropcms.org/forum/how-do-i-define-weight-block-layout

**Comment by kiamlaluno** [comment]
IMO, the weight value is relevant only on a list of nodes / taxonomy terms / links / etc. It does not make sense to change the weight value when editing a single item, especially when the same item could be part of different lists.

I would rather remove the form element for the weight value, instead of hiding it.
